frame/recovery/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+				.ok_or(Error::<T>::Overflow)?;
+			let total_deposit = T::ConfigDepositBase::get()
+				.checked_add(&friend_deposit)
+				.ok_or(Error::<T>::Overflow)?;
+			// Reserve the deposit
+			T::Currency::reserve(&who, total_deposit)?;
+
+			// Create the recovery configuration
+			let recovery_config = RecoveryConfig {
+				delay_period,
+				deposit: total_deposit,
+				friends,
+				threshold,
+			};
+
+			<Recoverable<T>>::insert(&who, recovery_config);
+
+			Self::deposit_event(RawEvent::RecoveryCreated(who));
+		}
+		
+		/// Allow a user to start the process for recovering an account.
+		fn initiate_recovery(origin, account: T::AccountId) {
+			let who = ensure_signed(origin)?;
+			// Check that the account is recoverable
+			ensure!(<Recoverable<T>>::exists(&account), Error::<T>::NotRecoverable);
+			// Check that the recovery process has not already been started
+			ensure!(!<ActiveRecoveries<T>>::exists(&account, &who), Error::<T>::AlreadyStarted);
+			// Take recovery deposit
+			let recovery_deposit = T::RecoveryDeposit::get();
+			T::Currency::reserve(&who, recovery_deposit)?;
+			// Create an active recovery status
+			let recovery_status = ActiveRecovery {
+				created: <system::Module<T>>::block_number(),
+				deposit: recovery_deposit,
+				friends: vec![],
+			};
+
+			<ActiveRecoveries<T>>::insert(&account, &who, recovery_status);
+
+			Self::deposit_event(RawEvent::RecoveryInitiated(account, who));
+		}
+		
+		/// Allow a friend to vouch for an active recovery process.
+		fn vouch_recovery(origin, lost: T::AccountId, rescuer: T::AccountId) {
+			let who = ensure_signed(origin)?;
+
+			// Get the recovery configuration for the lost account.
+			let recovery_config = Self::recovery_config(&lost).ok_or(Error::<T>::NotRecoverable)?;
+			// Get the active recovery process for the rescuer.
+			let mut active_recovery = Self::active_recovery(&lost, &rescuer).ok_or(Error::<T>::NotStarted)?;
+			// Make sure the voter is a friend
+			ensure!(Self::is_friend(&recovery_config.friends, &who), Error::<T>::NotFriend);
+
+			// Either insert the vouch, or return an error that the user already vouched.
+			match active_recovery.friends.binary_search(&who) {
+				Ok(_pos) => Err(Error::<T>::AlreadyVouched)?,
+				Err(pos) => active_recovery.friends.insert(pos, who.clone()),
+			}
+
+			// Update storage with the latest details
+			<ActiveRecoveries<T>>::insert(&lost, &rescuer, active_recovery);
+
+			Self::deposit_event(RawEvent::RecoveryVouched(lost, rescuer, who));
+		}
+		
+		/// Allow a rescuer to claim their recovered account.
+		fn claim_recovery(origin, account: T::AccountId) {
+			let who = ensure_signed(origin)?;
+			// Get the recovery configuration for the lost account
+			let recovery_config = Self::recovery_config(&account).ok_or(Error::<T>::NotRecoverable)?;
+			// Get the active recovery process for the rescuer
+			let active_recovery = Self::active_recovery(&account, &who).ok_or(Error::<T>::NotStarted)?;
```

Comment:
There is a possible optimization that can be done with the `claim_recovery` function.

You can see here that we read from storage both the `recovery_config` storage item and the `active_recovery` storage item.

We only need the `recovery_config` storage item to get the `threshold` value and check that it is less than or equal to the number of friends stored in the `active_recovery` storage item.

The optimization would be to copy the `threshold: u16` value into the `active_recovery` struct so that we don't need to read from the `recovery_config`.

So which is more efficient?

* One additional storage read per call to `claim_recovery`.
* One additional `u16` value per `active_recovery`. (this means across all active recoveries across all recoverable accounts).

Reply:
My guess is that duplicating the u16 value is better overall, and I will likely make an update to fix this.